upcoming/composable-pipelines-pattern.md

```diff
@@ -0,0 +1,392 @@
+def fasta_reader(filename):
+	with open(filename,'r') as infile:
+		for line in infile:
+			yield line
+
+# A base complementer, which converts each nucleotide in a
+# DNA sequence into its "complement" nucleotide (the one it
+# will sit paired to in the double-strand DNA helix molecule)
+def base_complementer(line_generator):
+	for line in line_generator:
+		if line[0] != '>':
+			translation_table = {
+			  'A' : 'T',
+			  'T' : 'A',
+			  'C' : 'G',
+			  'G' : 'C',
+			  'N' : 'N',
+			  '\n' : '\n'}
+			for nuc in line:
+				newline.append(translation_table[nuc])
+			yield newline
+
+def main():
+	# Connect our super-minimal little "workflow" consisting
+	# - a base complementer
+	# - a file reader
+	# - our print statements right here in the loop
+	fa_reader = fasta_reader('chr_y.fa')
+	for line in base_complementer(fa_reader):
+		print line
+````
+
+Have a careful look at this line:
+
+````python
+for line in base_complementer(fa_reader):
+	...
+````
+
+We here see how, rather than sending data into `base_complementer`, we instead send a generator object (`fa_reader`), that `base_complementer` will iterate over, as soon as we start iterating over the the latter.
+
+So when executing the main loop in `main()``, it will step by step drive our little pipeline consisting of three parts; a (FASTA) reader, a base complementer generator function and a print-statement at the end.
+
+Again, the lazy evaluation means that one item at a time will be drawn through the whole pipeline for each iteration in the main loop, without any temporary aggregates (lists or dicts) of lines building up between the components, which means the program needs to use very little memory during the process.
+
+## The Generator pattern in Go
+
+Coming into Go, the author was highly intrigued by all the powerful concurrency patterns made possible in this language, elaborated in blog posts such as [the one on "concurrency patterns"](http://blog.golang.org/pipelines) and ["advanced concurrency patterns"](http://blog.golang.org/advanced-go-concurrency-patterns).
+
+Even more interestingly, the simple and straight-forward generator pattern in python was easy to implement in Go too, as shown in [Rob Pike's Go Concurrency pattern slides](https://talks.golang.org/2012/concurrency.slide#25), and also [listed on this site](http://www.golangpatterns.info/concurrency/generators).
+
+The above python code would read something like this in Go, using the generator pattern (leaving out some imports and const definitions, for brevity):
+
+````go
+
+// ...<snip>...
+
+func fileReaderGen(filename string) chan []byte {
+	fileReadChan := make(chan []byte, BUFSIZE)
+	go func() {
+		file, err := os.Open(filename)
+		if err != nil {
+			log.Fatal(err)
+		} else {
+				scan := bufio.NewScanner(file)
+				for scan.Scan() {
+					// Write to the channel we will return, while copying
+					// the buffer slice (which is re-used) ...
+					fileReadChan <- append([]byte(nil), scan.Bytes()...)
```

Comment:
scan.Bytes() will reuse the slice it returns, as it progresses with it's scanning, so in order not to get garbled data when passing it on to another go-routine, the content has to be stuffed away into a new empty slice. This is the solution [I was proposed by Dave, Rob & Co](https://groups.google.com/forum/#!searchin/golang-nuts/samuel$20lampa$20scan.bytes/golang-nuts/5elNYb5J-aI/E--fwhcQJYAJ) in the mailing list. 

Have added clearer comments though, everywhere this occurs!